spec.bs

```diff
@@ -974,6 +987,29 @@ navigables</a> section, add the following:
         1. Set |newPageBudget|'s [=overall budget=] to [=overall page entropy allowance=].
         1. Set |navigable|'s [=page budget=] to |newPageBudget|.
 
+  #### Saved queries #### {#saved-queries}
+
+  <div algorithm>
+    To <dfn>get the index for a saved query</dfn>, given [=/navigable=] |navigable|, [=url/origin=] |origin|, [=string=] |operationName|, and [=string=] |savedQueryName|:
+
+    1. Let |topLevelTraversable| be the result of running [=get the top-level traversable=] for |navigable|.
+    1. [=Assert=] that |topLevelTraversable|'s [=page budget=] is not null.
+    1. If |topLevelTraversable|'s [=page budget=]'s [=saved query map=] does not [=map/contain=] |origin|, then return null.
+    1. If |topLevelTraversable|'s [=page budget=]'s [=saved query map=][|origin|] does not [=map/contain=] |operationName|, then return null.
+    1. If |topLevelTraversable|'s [=page budget=]'s [=saved query map=][|origin|][|operationName|] does not [=map/contain=] |savedQueryName|, then return null.
+    1. Return |topLevelTraversable|'s [=page budget=]'s [=saved query map=][|origin|][|operationName|][|savedQueryName|].
+  </div>
+
+  <div algorithm>
+    To <dfn>store the index for a saved query</dfn>, given [=/navigable=] |navigable|, [=url/origin=] |origin|, [=string=] |operationName|, [=string=] |savedQueryName|, and {{unsigned long}} |index|:
+
+    1. Let |topLevelTraversable| be the result of running [=get the top-level traversable=] for |navigable|.
+    1. [=Assert=] that |topLevelTraversable|'s [=page budget=] is not null.
+    1. If |topLevelTraversable|'s [=page budget=]'s [=saved query map=] does not [=map/contain=] |origin|, then [=map/set=] |topLevelTraversable|'s [=page budget=]'s [=saved query map=][|origin|] to an [=map/empty=] [=map=].
```

Comment:
Rather than a map containing a map that contains a further map, you can instead use a tuple as the map key.  For example:

```
[=map/Set=] |topLevelTraversable|'s [=page budget=]'s [=saved query map=][(|origin|, |operationName|, |savedQueryName|)] to |index|.
```

Reply:
An example of tuples as keys here: https://html.spec.whatwg.org/multipage/webappapis.html#module-map

Reply:
Done.

Reply:
Done.

Reply:
LGTM